### DISCUSSION TITLE
How to install Poetry on Windows so that it becomes portable?

### DISCUSSION THREAD

**Discussion by stefaneidelloth** [python-poetry/poetry#7035] [Q&A]
Hello, 

I'm not sure if this is a bug or a documentation issue. Please feel free to move the issue accordingly if its only about documentation.

**A.** If I install poetry with **pip on windows**, I am able to use poetry. 
`pip install poetry`

However, if I move my **portable python instance** to another folder, poetry cannot find python any more:

Fatal error in launcher: Unable to create process using '"**C:\python_env**\App\WinPython\python-3.10.5.amd64\python.exe"  "**D:\eis\python_env**\App\WinPython\python-3.10.5
.amd64\Scripts\poetry.exe" ': The system cannot find the file specified.

=> Poetry seems to store an absolute path to python.exe and that causes issues. 

(Other python packages, e.g. pip, work fine in portable WinPython).

**B.** In the poetry documentation the **recommend way to install poetry on windows** is to use a **PowerShell script**.

https://python-poetry.org/docs/#installation

=> How can I **modify the installation command** for Windows to tell poetry to use a **relative path** to python?

I tried following:

```
PS D:\eis\python_env> (Invoke-WebRequest -Uri https://install.python-poetry.org -UseBasicParsing).Content | POETRY_HOME=./App/WinPython/python-3.10.5.amd64/Scripts |
 py -
```

However, the position for POETRY_HOME does not seem to be correct:

```
The term 'POETRY_HOME=./App/WinPython/python-3.10.5.amd64/Scripts' is not recognized as the name of a  
cmdlet, function, script file, or operable program.
```

 And even if it would work, I don't know if poetry would use an absolute or relative path for python.

=> What is the correct installation command if my python instance is currently under

`D:\eis\python_env\App\WinPython\python-3.10.5.amd64`

and I want poetry to stay portable on Windows?

**C.** Maybe there is some sort of **poetry config file** or environment variable, where I could **modify some absolute path after installation** to be a relative path, so that poetry is portable? For example:

C:\python_env\App\WinPython\python-3.10.5.amd64\python.exe

=>

../../../../App/WinPython/python-3.10.5.amd64/python.exe

I searched inside the files of the package folder

D:\eis\python_env\App\WinPython\python-3.10.5.amd64\Lib\site-packages\poetry

for occurances of "WinPython" but could not find the right place to adapt the absolute path that is currently used.

I also tried to modify POETRY_HOME and POETRY_CONFIG_DIR but that did not help. 
=> Is there something like POETRY_PYTHON_HOME ?   (Could not find it in https://python-poetry.org/docs/configuration/ )

Or can I somehow tell poetry to use PYTHON_PATH instead of a hard coded, absolute installation path?

**Comment by stefaneidelloth** [comment]
Since poetry does not seem to be portable I'll switch back to pip as package manager.

**Comment by neersighted** [comment]
Poetry creates a standard virtual environment. There is no limitation here imposed by Poetry. I would suggest you simply create a virtual environment with `/path/to/python -m venv /path/to/poetry/venv` and `/path/to/poetry/venv/pip install poetry`.

I think you're getting stuck in the weeds on what is fundamentally simple, in any case. e.g. `curl ... | POETRY_HOME= py -` will only work on Unix, you cannot export environment variables inline in cmd/PowerShell. Set the environment variable in a step **before** running `curl | py` and things will work as you expect.

That being said, you needs are specialized, so exercising more control by performing the install manually is preferred.

**Comment by stefaneidelloth** [comment]
>There is no limitation here imposed by Poetry

Well, poetry.exe includes and absolute path to python.exe

**Reply by neersighted** [reply]
`poetry.exe` is created by `pip` and has no special needs. Poetry does not distribute it; Poetry is merely a Python package with an entrypoint. If you have tooling for relocatable virtual environments already, Poetry will just work as-is.

**Reply by stefaneidelloth** [reply]
Other python packages like jupyter, flask, ... also yield *.exe files in the Script folder but do not have the issue that they stop working if I move the folder. Therefore, I thought the issue is due to poetry. If there is a way to adapt the pip installation process, so that no absolute path is stored in the resulting *.exe file, please let me know. 
I do not use virtual environments. Main benefit of using poetry would be that I can include requirements in pyproject.toml

**Reply by neersighted** [reply]
Poetry is a package with dependencies, you should package it in a virtual environment if you do not strictly control your dependency versions as Linux distributions do (pip install poetry; pip install <otherpackage> in the same environment **will break** at unknown times on you).

Again, Poetry does not have any control over how it is installed; the tool you use to install it (probably pip) is what creates the executable wrapper script. If you are not the packager, but in fact an end user (that is not clear), you'll likely have to consult the distributor of your Python toolchain.

**Reply by slowglow** [reply]
OK, another poetry and WinPython end-user here (just making it clear). Do you know of any "tooling for relocatable virtual environments"? If you know, please share it with us. From what I tried, creating a venv with WinPython and installing poetry with `pip install` does not make a portable install of poetry. Same result if poetry was installed on the regular portable WinPython with the POETRY_HOME set up in my WinPython.ini file -> POETRY_HOME is used, but the *.exe inside the scripts dir is still with hard-coded paths, so it is immovable. If by "the  distributor of the toolchain" you mean WinPython, it has been contacted by the original poster and the issue is still open, but it seems the author does not know what to do about it (https://github.com/winpython/winpython/issues/1128). If by the toolchain distributor, you mean pypa, which stands behind pip and hosts packages, I don't think, they have been contacted about this particular issue. And, honestly, I don't think its their responsibility. As a simple user, I simply assume that packages uploaded to pypa, would simply work, and if not, it's up to the uploader to make them work after getting user feedback. So that's why this issue was brought here in the first place. As a user, I am dissapointed that this issue was closed and moved to a discussion, without offering any real solution and simultaneously preventing any of the other developers or would-be contributors to seeing it as an open issue in need of attention.

**Reply by stefaneidelloth** [reply]
For my use case I switched to hatch, being part of pypa: https://github.com/pypa/hatch
Also see https://stackoverflow.com/questions/74508024/is-requirements-txt-still-needed-when-using-pyproject-toml

**Reply by slowglow** [reply]
I am currently trying out PDM as an alternative and it seems to work. I still haven't moved my workflow to PDM, though. In a [blog post]( https://chriswarrick.com/blog/2023/01/15/how-to-improve-python-packaging), which had some strong (but well supported) opinions about PyPA's inability to steer the development of packaging tools, it was argued to be a good alternative.

Update: Unlike Poetry, which can install directly from github repos, I can't get PDM work for me in the same way. I get difficult to track down "The system cannot find the file specified" errors. I am not sure if I need to install git separately system-wise. I currently use a portable local install.

Update 2: Adding a portable version of git on the path, fixes the issue of PDM not being able to install from github. It also resolves the dependencies faster.

**Reply by slowglow** [reply]
Trying a slightly different method to install poetry on WinPython, produced (as expected) the same result.
This time I followed the "Manually (advanced)" installation instructions for Linux and modified them for Windows and WinPython in particular:
```
set VENV_PATH=%HOME%\my_venv
python -m venv %VENV_PATH%
%VENV_PATH%\Scripts\pip install -U pip setuptools
%VENV_PATH%\Scripts\pip install poetry
```
For WinPython installed in `NewProj_WPy64` the above commands are executed after opening the "WinPython Command Prompt.exe", which already has the environment variables pre-set for portability.

`my_venv\Include`, `my_venv\Lib`, `my_venv\Scripts` are created under `settings` (= `HOME` for WinPython).  `poetry.exe` is in `my_venv\Scripts`.

```
cd %HOME%
poetry new poetry_demo
```
creates a new project folder with all the files as expected.

After closing the command prompt, then renaming/moving the WinPython folder to `NewProj2_WPy64` poetry stops working.
Open a new "WinPython Command Prompt.exe" from the new location and try to add dependencies.

```
cd %HOME%\my_venv\poetry_demo
%HOME%\my_venv\Scripts\poetry add pendulum
```
This spits:
```
Fatal error in launcher: Unable to create process using '"D:\NewProj_WPy64\settings\my_venv\Scripts\python.exe"  "D:\NewProj2_WPy64\settings\my_venv\Scripts\poetry.exe" add pendulum': The system cannot find the file specified.
```
As you can see, the system cannot find the file specified, because old python location is hardcoded in the poetry executable.

**Comment by longlongvip** [comment]
Does it still need to be discussed? The officail document about **Advanced Install on Windows** shoule be modified

**Reply by slowglow** [reply]
I do not know how to answer that. 
Is there already an answer to the original question, besides using a different tool for managing dependencies?

**Reply by slowglow** [reply]
On python-poetry.org as of today (23-07-10) there is no section named "Advanced Install on Windows". -> It should be created first.

**Comment by alexchandel** [comment]
Can you please get `poetry` added to the `winget` package manager?

Installing poetry on Windows is a nightmare at the moment.  Not easy to guide someone else to install.
